keps/sig-storage/177-volume-snapshot/tighten-validation-webhook-crd.md

```diff
@@ -0,0 +1,871 @@
+3. User fixes their problems with invalid objects
+4. User upgrades the control plane again.
+5. In an n+2 release, once all the invalid objects are purged, we can switch the storage version to v1.
+
+In phase 2, the storage version will be kept at v1beta1 in order to ensure the rollback is possible.
+
+In phase 3, the storage version will be changed to v1.
+
+```yaml
+v1 (served=true, storage=false)
+v1beta1 (served=false, storage=true)
+```
+
+#### Current Controller validation of OneOf semantic
+
+##### Handling VolumeSnapshot.
+
+See code [here](https://github.com/kubernetes-csi/external-snapshotter/blob/v2.1.1/pkg/common-controller/snapshot_controller.go#L192).
+
+If the object violates oneOf semantic: Update the VolumeSnapshot status to “SnapshotValidationError” and issue an event.
+
+Note:
+- If the VolumeSnapshot object has been updated AFTER binding to a VSC, binding from VolumeSnapshot->VSC will be lost.
+- Deletion of an invalid resource is not blocked by that check as the deletion workflow happens before validation(code). This is to ensure that a user can delete an invalid VolumeSnapshot resource.
+
+##### Handling VolumeSnapshotContent
+
+See code [here](https://github.com/kubernetes-csi/external-snapshotter/blob/v2.1.1/pkg/common-controller/snapshot_controller.go#L91).
+If the object violates oneOf semantic: Update the VSC status to “ContentValidationError” and issue an event.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Deployment
+
+There are two main steps to setup validation for the snapshot objects. The kubernetes API server must be configured to connect to the webhook server, and the webhook server must be deployed and reachable. Make sure to take a look at the [prerequisites](#background-on-admission-webhooks) before deploying.
+
+A sample script will be provided which will handle the deployment of TLS certificates. It is not considered production ready and users are encouraged to use their own certificate management process. The demo will create certificates as a secret in the cluster and mount them as a volume. The `ValidatingWebhookConfiguration` will need to be updated with the cluster admin bundle.
+
+### Kubernetes API Server Configuration
+
+The API server must be configured to connect to the webhook server for certain API requests. This is done by creating a ValidatingWebhookConfiguration object. For a more thorough explanation of each field refer to the documentation. An example yaml file is provided below. The value of timeoutSeconds will affect the latency of snapshot creation, and must be considered carefully as it will affect the time the application may be frozen for.
```

Comment:
"it **will** affect the time the application may be frozen for." -> "it **might** affect the time the application may be frozen for."